[refactor] replace lazy_static with LazyLock

cylewitruk opened the pull request:
Resolves #5035 however using `std::sync::LazyLock` instead of `std::cell::LazyCell` because most of the values are used as statics (`LazyCell` is for `const`-like behavior). Also changed a bunch of `.unwrap()`s to `.expect()`s to improve eventual error interpretation.

`const` + `LazyCell` could likely be used for a handful of the values, but to determine those the usage of each needs to be evaluated, so I opted for `LazyLock` for all of them as it most closely mimics `static_lazy!`.

cylewitruk committed:
fmt-stacks

cylewitruk commented on testnet/stacks-node/Cargo.toml:
```diff
@@ -39,7 +38,7 @@ tikv-jemallocator = {workspace = true}
 ring = "0.16.19"
 warp = "0.3.5"
-tokio = "1.15"
+tokio = { version = "1.15", features = ["rt-multi-thread"] }
```
`Runtime::new()` only exists when this feature is enabled, the tests wouldn't compile for me otherwise, hence why I added the `rt-multi-thread` feature.

cylewitruk committed:
Merge branch 'develop' into chore/replace-lazy-static-with-lazy-lock

jbencin commented on clarity/src/vm/ast/parser/v1.rs:
```diff
@@ -112,26 +112,32 @@ fn get_lines_at(input: &str) -> Vec<usize> {
+pub static STANDARD_PRINCIPAL_REGEX: LazyLock<String> =
+    LazyLock::new(|| "[0123456789ABCDEFGHJKMNPQRSTVWXYZ]{28,41}".into());
+pub static CONTRACT_NAME_REGEX: LazyLock<String> = LazyLock::new(|| {
+    format!(
```
I think it would be better to use the `const_format` crate ([link](https://crates.io/crates/const_format)) for format strings. It's cleaner and results in  constant `&str`s that don't require locking to access

cylewitruk replied:
I'm not familiar with that crate, but this PR is only refactoring away `lazy_static` and not changing anything else.

How would you use that crate here? The highlighted code is constructing a `Regex` static lazily, not formatting.

cylewitruk replied:
Ah nevermind, I'm looking on my phone and it was a bit hard to see -- I think I see what you mean, I'll try it.

jbencin replied:
Adding `const_format` could be done in a separate PR, but there are places where the use of either `lazy_static!()` or `LazyLock` is unnecessary, and that should be addressed in this PR

cylewitruk replied:
Yeah, my goal was to simply remove the lazy_static dep with as little change as possible, but i'm happy to look at further refactorings. Will go through all your comments this weekend 👍

jbencin commented on stacks-common/src/util/log.rs:
```diff
@@ -258,9 +257,7 @@ fn inner_get_loglevel() -> slog::Level {
-lazy_static! {
-    static ref LOGLEVEL: slog::Level = inner_get_loglevel();
-}
+static LOGLEVEL: LazyLock<slog::Level> = LazyLock::new(|| inner_get_loglevel());
```
Can probably simplify to

```suggestion
static LOGLEVEL: LazyLock<slog::Level> = LazyLock::new(inner_get_loglevel);
```

jbencin commented on stackslib/src/chainstate/coordinator/tests.rs:
```diff
@@ -71,12 +70,12 @@ use crate::util_lib::boot::{boot_code_addr, boot_code_id};
-    pub static ref MBLOCK_PUBKHS: Arc<AtomicU64> = Arc::new(AtomicU64::new(1));
-    pub static ref STACKS_BLOCK_HEADERS: Arc<AtomicU64> = Arc::new(AtomicU64::new(1));
-}
+pub static BURN_BLOCK_HEADERS: LazyLock<Arc<AtomicU64>> =
```
It seems like we could just use `AtomicU64` here without the extra wrappers

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -148,127 +148,137 @@ impl FromSql for HeaderTypeNames {
+]
+});
+
+pub static NAKAMOTO_CHAINSTATE_SCHEMA_2: LazyLock<Vec<String>> = LazyLock::new(|| {
```
There are no format strings here, it could just be a `[&'static str; 3]`

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -148,127 +148,137 @@ impl FromSql for HeaderTypeNames {
+    ]
+});
+
+pub static NAKAMOTO_CHAINSTATE_SCHEMA_3: LazyLock<Vec<String>> = LazyLock::new(|| {
```
There are no format strings here, it could just be a `[&'static str; 2]`

jbencin commented on stackslib/src/chainstate/stacks/boot/mod.rs:
```diff
@@ -108,38 +108,52 @@ pub const MINERS_NAME: &'static str = "miners";
+pub static POX_4_CODE: LazyLock<String> = LazyLock::new(|| POX_4_BODY.to_string());
+
+pub static BOOT_CODE_COST_VOTING_TESTNET: LazyLock<String> =
+    LazyLock::new(|| make_testnet_cost_voting());
```
```suggestion
    LazyLock::new(make_testnet_cost_voting);
```

jbencin commented on stackslib/src/monitoring/mod.rs:
```diff
@@ -38,9 +37,8 @@ use crate::util_lib::db::{sqlite_open, tx_busy_handler, DBConn, Error as Databas
-lazy_static! {
-    static ref GLOBAL_BURNCHAIN_SIGNER: Mutex<Option<BurnchainSigner>> = Mutex::new(None);
-}
+static GLOBAL_BURNCHAIN_SIGNER: LazyLock<Mutex<Option<BurnchainSigner>>> =
```
Rust has allowed static mutexes for a couple years now, it's not necessary to put it in a `LazyLock`

jbencin commented:
Don't forget to update `package.rust-version` to `1.80` in Cargo.toml files where `LazyLock` is used

cylewitruk committed:
wip: further refactorings for lazy_static/lazycell/lazylock

cylewitruk committed:
Merge remote-tracking branch 'refs/remotes/origin/chore/replace-lazy-static-with-lazy-lock' into chore/replace-lazy-static-with-lazy-lock

cylewitruk committed:
refactor unnecessary 'static's and cargo fmt

cylewitruk commented:
@jbencin I think all of your comments are resolved now. Really like `const_format`, good tip!

I went a step further and reverted to `const LazyCell` where appropriate, changed a bunch of `static`s to `const`s which didn't need to be static and did some other general cleanup.

jbencin commented on Cargo.toml:
```diff
@@ -1,5 +1,6 @@
 [workspace]
 resolver = "2"
+package.rust-version = "1.80"
```
I don't know about using a workspace-level MSRV... it's fine to require latest version for `stacks-node`, but we might want to be more conservative with stacks-common and stackslib, as those are used by other projects

cylewitruk replied:
I thought about that too, but now when `clarity`, `stacks-common` and `stackslib` all use `LazyCell` and/or `LazyLock`, they need to be at 1.80 in any case. We can keep the `rust-version` flag local to the packages if that's preferred, I just figured it's easier to keep the whole project at an announced MSRV :shrug: No strong opinions here.

jbencin replied:
While it's technically correct because the MSRV is 1.80 for all packages right now, and even though it can be overriden in package-level Cargo.toml files, this implies that we have some global MSRV policy which applies to all packages, which I don't think we want. So IMO it's better not to do this, but I don't really have strong feelings on it, and I'd approve either way

jcnelson replied:
@jbencin @obycode who need to weigh in here?  Should the ClarityWASM folks and Clarinet folks weigh in? cc @hugocaillard

jbencin replied:
@jcnelson Hugo's on vacation right now, but I checked and Clarinet builds on Rust 1.80 without warnings. There are commit messages like this in the history:

```
445ea332 (4 weeks ago) Hugo C: fix: clippy warning v1.80 (#1518)
```

So this PR shouldn't be a problem for them

jcnelson replied:
Alright, it's fine by me as long as its easy to change back if someone complains.

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -148,127 +149,125 @@ impl FromSql for HeaderTypeNames {
 
-    pub static ref NAKAMOTO_CHAINSTATE_SCHEMA_2: Vec<String> = vec![
-    NAKAMOTO_TENURES_SCHEMA_2.into(),
+pub static NAKAMOTO_CHAINSTATE_SCHEMA_2: &[&str] = &[
```
Can be `const` I think

```suggestion
pub const NAKAMOTO_CHAINSTATE_SCHEMA_2: &[&str] = &[
```

cylewitruk replied:
Yeah, missed it, good catch

jbencin commented on stackslib/src/chainstate/nakamoto/mod.rs:
```diff
@@ -148,127 +149,125 @@ impl FromSql for HeaderTypeNames {
 
-    pub static ref NAKAMOTO_CHAINSTATE_SCHEMA_3: Vec<String> = vec![
-    NAKAMOTO_TENURES_SCHEMA_3.into(),
+pub static NAKAMOTO_CHAINSTATE_SCHEMA_3: &[&str] = &[
```
```suggestion
pub const NAKAMOTO_CHAINSTATE_SCHEMA_3: &[&str] = &[
```

cylewitruk committed:
chore: update two missed static->const refactors

cylewitruk committed:
Merge branch 'develop' into chore/replace-lazy-static-with-lazy-lock

obycode reviewed (approved):
lgtm!

jcnelson commented on Cargo.lock:
```diff
@@ -4,9 +4,9 @@ version = 3
+version = "0.22.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "8a30b2e23b9e17a9f90641c7ab1549cd9b44f296d3ccbf309d2863cfe398a0cb"
+checksum = "6e4503c46a5c0c7844e948c9a4d6acd9f50cccb4de1c48eb9e291ea17470c678"
```
What's with the bump?

jcnelson reviewed (changes requested):
Why are there so many bumps in the cargo.lock?

jcnelson commented on Cargo.lock:
```diff
@@ -1859,15 +1884,15 @@ dependencies = [
+version = "1.5.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "e2abad23fbc42b3700f2f279844dc832adb2b2eb069b2df918f455c4e18cc646"
+checksum = "bbd2bcb4c963f2ddae06a2efc7e9f3591312473c50c6685e1f298068316e66fe"
```
I thought we were dropping this?

jbencin replied:
It's still used by some of our dependencies though, like `prometheus`
